lms/views/predicates/_lti_launch.py

```diff
@@ -41,13 +41,21 @@ def __call__(self, context, request):
         resource_link_id = request.params.get("resource_link_id")
         tool_consumer_instance_guid = request.params.get("tool_consumer_instance_guid")
 
-        has_module_item_configuration = (
-            ModuleItemConfiguration.get_document_url(
-                request.db, tool_consumer_instance_guid, resource_link_id
-            )
-            is not None
+        document_url = ModuleItemConfiguration.get_document_url(
+            request.db, tool_consumer_instance_guid, resource_link_id
         )
 
+        if document_url:
+            has_module_item_configuration = True
+        else:
+            previous_resource_link_id = request.params.get("resource_link_id_history")
+
+            previous_document_url = ModuleItemConfiguration.get_document_url(
+                request.db, tool_consumer_instance_guid, previous_resource_link_id
+            )
+
+            has_module_item_configuration = bool(previous_document_url)
```

Comment:
`get_document_url()` gets called four times in total during a DB-configured launch (in the worst case). I think the solution is to add caching to `get_document_url()` though, rather than complicate the code by trying to pass its return value around. I'd have to think about caching SQLAlchemy query results and how that might interact with DB transactions. Also it might be that SQLAlchemy is already caching this for us. An optimisation to look into for in a follow-up PR.

Reply:
So the end result of this code is that a view with `@view_config(db_configured=True)` will only be called if:

* The request has a `resource_link_id` param and we do have a `ModuleItemConfiguration` for it in the DB
* Or, failing that, the request has a `resource_link_id_history` param and we do have a `ModuleItemConfiguration` for that in the DB

The view will not be called if there isn't a `ModuleItemConfiguration` matching either the `resource_link_id` or the `resource_link_id_history` in the DB, even if both params are in the request.